crane/data.py

```diff
@@ -79,18 +79,41 @@ def monitor_data_dir(app, last_modified=0):
     if not os.path.exists(data_dir):
         logger.error('The data directory specified does not exist: %s' % data_dir)
     while True:
-        # Check if the modified time has changed on the directory and if so reload the data
-        # This has been verified using a directory mounted via NFS 4 as well as locally
-        try:
-            logger.debug('Checking for new metadata files')
-            current_modified = os.stat(data_dir).st_mtime
-            if current_modified > last_modified:
-                last_modified = current_modified
-                load_all(app)
-        except OSError:
-            # The directory does not exist
-            pass
-        time.sleep(polling_interval)
+        # Find all the subdirectories
+        dirs = [dirpath for (dirpath, _, _) in os.walk(data_dir,
+                                                       followlinks=True)]
+
+        # Find the most recent mtime
+        if dirs and not last_modified:
+            try:
+                last_modified = max(os.stat(dir).st_mtime for dir in dirs)
+            except OSError:
+                # One of the directories no longer exists since the
+                # os.walk() call above. Load everything, sleep, and
+                # try again.
+                pass
+
+        # Load everything
+        load_all(app)
+
+        # Wait for something to change
+        while True:
+            time.sleep(polling_interval)
+            if not dirs:
+                # The data directory didn't exist before so re-check
+                break
+
+            # Check if the modified time on any directory is newer
+            # than the most recent change we saw
+            try:
+                logger.debug('Checking for new metadata files')
+                most_recent = max(os.stat(dir).st_mtime for dir in dirs)
```

Comment:
A DRY improvement might be to make this max() thing a function, even though it's a one liner.